src/cleo/terminal.py

```diff
@@ -0,0 +1,67 @@
+from __future__ import annotations
+
+import os
+import sys
+
+from typing import NamedTuple
+
+
+class TerminalSize(NamedTuple):
+    width: int
+    height: int
+
+
+class Terminal:
+    def __init__(
+        self,
+        width: int | None = None,
+        height: int | None = None,
+        fallback: tuple[int, int] | None = None,
+    ) -> None:
+        self._width = width
+        self._height = height
+        self._fallback = TerminalSize(*(fallback if fallback is not None else (80, 25)))
+
+    @property
+    def width(self) -> int:
+        return self.size.width
+
+    @property
+    def height(self) -> int:
+        return self.size.height
+
+    @property
+    def size(self) -> TerminalSize:
+        return self._get_terminal_size()
+
+    def _get_terminal_size(self) -> TerminalSize:
+        if self._width is not None and self._height is not None:
+            return TerminalSize(self._width, self._height)
+
+        width = 0
+        height = 0
+
+        columns = os.environ.get("COLUMNS")
+        if columns is not None and columns.isdigit():
+            width = int(columns)
+        lines = os.environ.get("LINES")
+        if lines is not None and lines.isdigit():
+            height = int(lines)
+
+        if width <= 0 or height <= 0:
+            try:
+                os_size = os.get_terminal_size(sys.__stdout__.fileno())
+                size = TerminalSize(*os_size)
+            except (AttributeError, ValueError, OSError):
+                # stdout is None, closed, detached, or not a terminal, or
+                # os.get_terminal_size() is unsupported # noqa: ERA001
+                size = self._fallback
+            if width <= 0:
```

Comment:
Does this make sense? If `os.get_terminal_size()` fails, it feels like we should treat the value as garbage and just return the fallback verbatim.

Reply:
This is exactly what is being done. Notice, that `os.get_terminal_size` could return `(0, 0)`, which would not be something we want, hence line 60 and 62.

Reply:
Right, but the logic here still seems slightly off to me -- what is setting width or height inside this conditional block? Do you mean to check `size.width` and `size.height`?

Reply:
no, `size` properties are checked inline in lines 60 and 62. These conditions are completing the checks started in line 51 (general check if any of 2 properties need to be filled with `os` function or fallback, and 59/61 are checks for which of those 2 need to be filled). In general, this code is taken from `shutil.get_terminal_size` function and fitted to our needs.